enabled AT_USE_JITERATOR() for `tan` and `tanh` kernels.

parth-desai opened the pull request:
This MR fixes the test failures for `jiterator` implemenation for `tan` and `tanh` Unary kernels as mentioned in #100842.

The failures were fixed by adjusting tolerances but some failures were in `test_unary_ufuncs.py` required adjusting input values as well. Since the jiterator kernels are using libstdc++, the supported input range is smaller than thrust implementation.

parth-desai committed:
enabled AT_USE_JITERATOR() for unary geometric kernels

parth-desai committed:
moved optoleranceoverride implementation to test_common_methods_invocations.py

parth-desai committed:
fixed tests in test_foreach.py and test_unary_ufuncs.py

parth-desai committed:
Added tolerance decorator to opinfo

parth-desai commented on test/test_foreach.py:
```diff
@@ -399,6 +399,16 @@ def _inplace_unary_test(self, inplace, inplace_ref, inputs, is_fastpath, **kwarg
 
     @ops(foreach_unary_op_db)
     @parametrize("is_fastpath", (True, False))
+    @opsToleranceOverride(foreach_unary_op_db, 'TestForeach', 'test_unary_op', (
```
Thanks for the pointer. I have updated the code based on your suggestion.

parth-desai committed:
fixed import

parth-desai committed:
fixed failing test in test_unary_ufuncs.py

parth-desai commented:
@mruberry @peterbell10 @ngimel Can you let me know if i need to format my code or update documentation anywhere for this change?

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9181,7 +9189,11 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+                               DecorateInfo(toleranceOverride({torch.complex64: tol(atol=1e-05, rtol=1e-05)}),
+                                            'TestUnaryUfuncs',
+                                            'test_reference_numerics_extremal',
+                                            device_type='cuda'),),
```
Lint is failing, see the [contributing guide](https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#local-linting) on how to run lint locally.

peterbell10 reviewed (commented):
The test failures look real, e.g.
```
test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_asinh_cuda_complex64 - AssertionError: Tensor-likes are not close!
```

parth-desai committed:
fixed formatting

parth-desai committed:
fixed failing test

parth-desai committed:
reverted tanhshrink related change

parth-desai committed:
Revert "reverted tanhshrink related change"

This reverts commit 074ab49069cedcd1edcb6f732904d9edc07650c3.

parth-desai commented:
@peterbell10 Could you take a look at this again?

peterbell10 commented:
@parth-desai would you mind resolving the merge conflicts? I'm hoping a rebase would resolve the CI issues.

parth-desai commented:
> @parth-desai would you mind resolving the merge conflicts? I'm hoping a rebase would resolve the CI issues.

@peterbell10 I merged the latest main to my branch. Please review again.

peterbell10 commented:
This test failure looks related:
```
2023-06-19T15:19:25.1218162Z =================================== FAILURES ===================================
2023-06-19T15:19:25.1218410Z _ TestUnaryUfuncsCUDA.test_reference_numerics_extremal_nn_functional_tanhshrink_cuda_bfloat16 _
2023-06-19T15:19:25.1218592Z Traceback (most recent call last):
2023-06-19T15:19:25.1218848Z   File "/var/lib/jenkins/workspace/test/test_unary_ufuncs.py", line 315, in test_reference_numerics_extremal
2023-06-19T15:19:25.1219016Z     self._test_reference_numerics(dtype, op, tensors)
2023-06-19T15:19:25.1219246Z   File "/var/lib/jenkins/workspace/test/test_unary_ufuncs.py", line 260, in _test_reference_numerics
2023-06-19T15:19:25.1219375Z     _helper_reference_numerics(
2023-06-19T15:19:25.1219613Z   File "/var/lib/jenkins/workspace/test/test_unary_ufuncs.py", line 206, in _helper_reference_numerics
2023-06-19T15:19:25.1219749Z     self.assertEqualHelper(
2023-06-19T15:19:25.1219974Z   File "/var/lib/jenkins/workspace/test/test_unary_ufuncs.py", line 171, in assertEqualHelper
2023-06-19T15:19:25.1220077Z     self.assertEqual(
2023-06-19T15:19:25.1220472Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 3099, in assertEqual
2023-06-19T15:19:25.1220610Z     raise error_metas[0].to_error(
2023-06-19T15:19:25.1220830Z AssertionError: Tensor-likes are not close!
2023-06-19T15:19:25.1220854Z 
2023-06-19T15:19:25.1220993Z Mismatched elements: 565 / 16384 (3.4%)
```
It's strange though, because this PR should only effect complex dtypes, but the test is for `bfloat16`.

parth-desai commented:
@peterbell10 I couldn't reproduce the error on my local machine. I even tried the CI docker image as well as build scripts. Any suggestions on fixing this test?

parth-desai committed:
fixed linting issues

peterbell10 commented:
@pytorchbot rebase

peterbell10 commented on aten/src/ATen/native/cuda/UnaryGeometricAcosKernel.cu:
```diff
@@ -12,14 +12,14 @@
 void acos_kernel_cuda(TensorIteratorBase& iter) {
   auto common_dtype = iter.common_dtype();
   if (at::isComplexType(common_dtype)) {
     // Disabled due to accuracy issues
```
You should remove these comments as well.

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -13441,6 +13498,9 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                          'TestCudaFuserOpInfo', 'test_nvfuser_correctness'),
+            DecorateInfo(toleranceOverride({torch.complex64: tol(atol=6e-04, rtol=1e-05)}),
+                         'TestUnaryUfuncs', 'test_reference_numerics_extremal',
+                         device_type='cuda'),
```
The bfloat16 failure is due to this `toleranceOverride` taking precedence over the one above for bfloat16. I can get the test to pass locally by adding copying bfloat16's `tol` term from above here.

peterbell10 reviewed (commented):
Just FYI, `main` often has some test failures from PRs that need to be reverted. You're best rebasing against `viable/strict`, which should be the most recent commit where CI passed. That should improve the signal-to-noise on CI.

I've set the `keep-going` flag so you should see all the tests failures, and it looks like there are still quite a lot of relevant CI failures for complex dtypes.

parth-desai committed:
fixed tanhshrink cuda bfloat16 failure

parth-desai committed:
removed stale comments

parth-desai commented:
@peterbell10 Thanks for the pointer. I didn't realize my environment had issues and my build was not running cuda tests. Anyways I have fixed the `bfloat16` issue and rebased with `viable/strict`. Could you run the CI again to see if there are any further failures?

parth-desai committed:
fixed test failures for large and small reference numerics

parth-desai commented:
@peterbell10 Can you review the changes again? I fixed the test failures but the changes I made for numerics filter were just to make jiterator happy. Something seems incorrect to me. 

I was trying to debug when failure were happening for large & small values I found that torch implementation return incorrect result while libc++ will return correct value. 

for example,
```
import torch
import numpy as np 

a = torch.tensor([-1e+19+1e+19j], dtype=torch.complex64, device='cuda')
print('torch asinh', a.asinh()) # will output torch asinh tensor([-inf+0.j], device='cuda:0')


x = np.array(a.cpu())
print('numpy asinh', np.arcsinh(x)) # will output numpy asinh [-44.788837+0.7853982j]
```

When I write a simple c++ implementation that is used for jiterator 
```
#include <complex>
#include <iostream>
 
int main()
{
    std::cout << std::fixed;
    std::complex<double> z1(-1e+19,1e+19);
    std::cout << "asinh" << z1 << " = " << std::asinh(z1) << '\n'; 
// output: asinh(-10000000000000000000.000000,10000000000000000000.000000) = (-44.788838,0.785398)
}

```

I am not sure why generated jiterator kernel returns incorrect result. I will spend more time on debugging it but any pointer you can provide would be great.

peterbell10 commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -1762,7 +1762,7 @@ def generate_elementwise_binary_large_value_tensors(
     _large_int_vals = (-1113, 1113, -10701, 10701)
     _large_float16_vals = (-501, 501, -1001.2, 1001.2, -13437.7, 13437.7)
-    _large_float_vals = _large_float16_vals + (-4988429.2, 4988429.2, -1e20, 1e20)
+    _large_float_vals = _large_float16_vals + (-4988429.2, 4988429.2, -1e19, 1e19)
```
These test sample filters seem really extreme considering the tests pass today without any of them. Perhaps you could investigate replacing the libc++ version we vendor with thrust's complex math library which has a permissive BSD license:
https://github.com/NVIDIA/thrust/tree/24486a169a62a58ef8f824d3dc9613c006b6f5a7/thrust/detail/complex

parth-desai replied:
@peterbell10 To confirm, You are suggesting to replace this implementation [llvm_complex.cpp](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/llvm_complex.cpp#L20) with thrust implementation?

Also Do you know of a good way to look at source code for generated JIT kernel, Currently I have to put `std::cout` in the execution path to see the source code. I wish to manually compile the generated source code using `nvcc`.

parth-desai committed:
Added thrust implementation

parth-desai committed:
moved to thrust impl

parth-desai commented:
@peterbell10 Can you please review the changes again? I have added thrust implementation. It is crude at the moment but it fixes the reference numerics test failures from earlier.

malfet reviewed (changes requested):
Please:
 - Rebase your PR against latest trunk
 - Provide an explanation what your change is trying to do
 - In particular, explain why it's necessary to vend thurst code in `llvm_complex.h`(aren't there a better location), but at the same time increase tolerances (as far as I understand from the original issues, it was suggested to do either one or another)
 - If thrust code must be vendored in, add a linter that check that it matches with thurst version PyTorch depends on (to avoid divergence if some bugs in embeded copy of trust will be fixed in newer releases)
 - Make sure PR passes CI
 
IMO a better approach would be to split this PR into several smaller PRs, as I assume `tan`/`tanh` indeed can be enabled by tweaking tolerance a bit, while inverse trigonometric functions perhaps require a more thorough investigation (as suggested in original issue)

parth-desai committed:
reverted thrust related changes

parth-desai committed:
reverted changes for asin and acos ops

parth-desai commented:
@malfet I have reduced the changes down to only `tan` and `tanh` kernels. Please review the changes again. I will open separate PRs for other kernels.

parth-desai commented:
@malfet The test failues seems to be unrelated to my changes. Any suggestions on how to fix or ignore them?

For example, [this](https://github.com/pytorch/pytorch/actions/runs/5934044826/job/16090707479#step:15:5314) test failure is related to [this](https://github.com/pytorch/pytorch/issues/106440) issue.

peterbell10 commented:
@pytorchbot rebase

parth-desai commented:
@pytorchbot rebase

parth-desai commented:
@peterbell10 The distributed test failures are also unrelated. I ran the tests locally and they are passing. I have rebased again to see if that would fix the failure.

huydhn commented:
@pytorchbot drci

huydhn commented:
@pytorchbot drci

peterbell10 commented:
These failures look real:
```
test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal__refs_tan_cuda_complex64
test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal__refs_tanh_cuda_complex64
test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal__refs_nn_functional_tanhshrink_cuda_complex64
```

parth-desai committed:
fixed failures with ref ops

parth-desai commented:
@peterbell10 I fixed the failures. Please run the CI.

parth-desai committed:
fixed tolerances

parth-desai commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -13682,6 +13704,9 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                 toleranceOverride({torch.bfloat16: tol(atol=1e-02, rtol=1.6e-02)}), 'TestUnaryUfuncs',),
             DecorateInfo(toleranceOverride({torch.float16: tol(atol=1e-02, rtol=1e-02)}),
                          'TestCudaFuserOpInfo', 'test_nvfuser_correctness'),
+            DecorateInfo(toleranceOverride({torch.complex64: tol(atol=4e-02, rtol=1e-0),
```
Good Catch! It was a typo. I have also fixed macOS failure.

parth-desai replied:
Nevermind on macOS failure. Not sure why macOS test is failing with Unexpected success. 
```
2023-08-30T03:45:46.9155230Z _ TestUnaryUfuncsCPU.test_reference_numerics_small__refs_nn_functional_tanhshrink_cpu_complex128 _
2023-08-30T03:45:46.9155280Z Unexpected success
```

peterbell10 replied:
The `reference_numerics_filter` is probably making the xfailed test pass, so you should be able to remove the `unittest.expectedFailure` decorator from the list of skips.

parth-desai committed:
removed expectedfailure for macos

parth-desai commented:
@peterbell10 The failures now seem to be CI infrastructure failures. Can you take a look?

huydhn commented:
@pytorchbot drci

huydhn commented:
@pytorchbot merge

parth-desai commented:
@huydhn Any suggestions on how to fix these?

huydhn commented:
@pytorchbot merge -f 'Bypass unrelated issues as Dr.CI has classified them as flaky'

Sorry I was about to force merge your change but got side tracked by other tasks

huydhn commented:
And thank you for your first contribution to PyTorch (IIRC) :)

parth-desai commented:
thank you @huydhn @peterbell10 for your help.
